### PR TITLE
[WIP] Add Point codec methods:

### PR DESCRIPTION
**Pull request by sublimator**
Add Point#eq
Add Point#encode
Add Point#encodeCompressed
Add Curve#decodePoint
Add overrides for Montgomery curves/points

Squash of #48

### CODE REVIEW

File: lib/elliptic/curve/base.js [hunk 1]
```diff
@@ -239,10 +240,51 @@ function BasePoint(curve, type) {
...
+
+  var len = this.p.byteLength();
+  if (bytes[0] === 0x04 && bytes.length - 1 === 2 * len) {
+    return this.point(bytes.slice(1, 1 + len),
+                      bytes.slice(1 + len, 1 + 2 * len));
+  } else if ((bytes[0] === 0x02 || bytes[0] === 0x03) &&
+              bytes.length - 1 === len) {
+    return this.pointFromX(bytes.slice(1, 1 + len), bytes[0] === 0x03);
+  }
+  throw new Error('Unknown point format');
```
Comment (sublimator) [inline lib/elliptic/curve/base.js]: Should this method try and check for validity of the point ?

Comment (indutny) [inline lib/elliptic/curve/base.js]: I wouldn't do it.

Comment (sublimator) [inline lib/elliptic/curve/base.js]: I think I agree

File: lib/elliptic/curve/mont.js [hunk 2]
```diff
@@ -147,6 +158,10 @@ Point.prototype.mulAdd = function mulAdd() {
   throw new Error('Not supported on Montgomery curve');
 };
 
+Point.prototype.eq = function eq(other) {
+  return this.getX().cmp(other.getX()) === 0;
```
Comment (sublimator) [inline lib/elliptic/curve/mont.js]: Does this need to check z is 1 or anything ?

Comment (indutny) [inline lib/elliptic/curve/mont.js]: To answer this question we need to asnwer: does `getX` normalize point?

Comment (sublimator) [inline lib/elliptic/curve/mont.js]: Indeed, I forgot about that :) And the normalized form puts z as 1 huh?

Comment (indutny) [inline lib/elliptic/curve/mont.js]: It is right below this function :)

Comment (sublimator) [inline lib/elliptic/curve/mont.js]: hahaha! That's funny

Comment (indutny) [inline lib/elliptic/curve/mont.js]: Its funny for you, but it scares me a lot. You patched this code all over, and didn't see a small function below. Now I will need to recheck everything :)

Comment (sublimator) [inline lib/elliptic/curve/mont.js]: Oh, come on, I was just thinking out loud, momentarily forgot the normalize
:)

Comment (sublimator) [inline lib/elliptic/curve/mont.js]: But yes, nothing wrong with rechecking :) :+1:

Comment (sublimator) [inline lib/elliptic/curve/mont.js]: If it calms you down, the `normalized form puts z as 1 huh?` was one of those questions that already had an answer.

Comment (indutny) [inline lib/elliptic/curve/mont.js]: Stay calm :) I was joking

Comment (sublimator) [inline lib/elliptic/curve/mont.js]: Keep calm and let @indutny handle it

**Comment by sublimator** [comment]
Anything else you want changed here, or you want it squashed down?

**Comment by sublimator** [comment]
@dcousens 
The party over here now :)

File: lib/elliptic/curve/base.js [hunk 3]
```diff
@@ -239,10 +240,51 @@ function BasePoint(curve, type) {
...
+BasePoint.prototype._encode = function _encode(compact) {
+  var len = this.curve.p.byteLength();
+  var x = intToBE(this.getX(), len);
+
+  return compact ?
+      [ this.getY().isEven() ? 0x02 : 0x03 ].concat(x) :
+      [ 0x04 ].concat(x, intToBE(this.getY(), len)) ;
+};
+
+BasePoint.prototype.encode = function encode(compact, enc) {
```
Comment (sublimator) [inline lib/elliptic/curve/base.js]: How do we feel about switching the order of the compact and enc args around ? In context of `encodeCompressed` ?

File: lib/elliptic/curve/base.js [hunk 4]
```diff
@@ -239,10 +240,46 @@ function BasePoint(curve, type) {
...
+  var len = this.curve.p.byteLength();
+  var x = intToBE(this.getX(), len);
+
+  if (compact)
+    return [ this.getY().isEven() ? 0x02 : 0x03 ].concat(x);
+
+  return [ 0x04 ].concat(x, intToBE(this.getY(), len)) ;
+};
+
+BasePoint.prototype.encode = function encode(enc, compact) {
```
Comment (sublimator) [inline lib/elliptic/curve/base.js]: I went ahead and swapped these around

File: lib/elliptic/utils.js [hunk 5]
```diff
@@ -170,10 +170,12 @@ function intFromLE(bytes) {
...
-  while (bytes.length < padTo)
-    bytes.push(0);
-  return bytes;
+function intToLE(num, length) {
+  return num.toArray('le', length);
 }
 utils.intToLE = intToLE;
+
+function intToBE(num, length) {
+  return num.toArray('be', length);
```
Comment (indutny) [inline lib/elliptic/utils.js]: Seems to be kind of small method, why not inline this code?

File: lib/elliptic/utils.js [hunk 6]
```diff
@@ -170,10 +170,12 @@ function intFromLE(bytes) {
 }
 utils.intFromLE = intFromLE;
 
-function intToLE(num, padTo) {
-  var bytes = num.toArray('le');
-  while (bytes.length < padTo)
-    bytes.push(0);
-  return bytes;
+function intToLE(num, length) {
+  return num.toArray('le', length);
```
Comment (indutny) [inline lib/elliptic/utils.js]: Same above. It is very slow, and not going to change much.

File: test/curve-test.js [hunk 7]
```diff
@@ -162,3 +162,87 @@ describe('Curve', function() {
...
+      var encoded = p.encode('hex');
+      var decoded = curve.decodePoint(encoded, 'hex');
+      assert(decoded.eq(p));
+      assert.equal(encoded, definition.encoded);
+    }
+  }
+
+  var shortPointEvenY = {
+    coordinates: {
+      x: "79be667ef9dcbbac55a06295ce870b07029bfcdb2dce28d959f2815b16f81798",
```
Comment (indutny) [inline test/curve-test.js]: Mixing `"` and `'`, let's use `'` everywhere.

File: lib/elliptic/utils.js (hunk 6 above)
Comment (sublimator) [inline lib/elliptic/utils.js]: Is there really any measurable performance cost?

File: test/curve-test.js (hunk 7 above)
Comment (sublimator) [inline test/curve-test.js]: Errrrk. Will use '

**Comment by sublimator** [comment]
@indutny 

I inlined those methods, and fixed the quoting issues, along with some other lint issues. It might pay to enable linting etc for the tests eh ( when moving to semi-standard probably a good time) ?

**Comment by indutny** [comment]
@sublimator there is a mental cost for them (at least for me). Linting tests makes sense to me!

**Comment by indutny** [comment]
LGTM, please squash ;)

**Comment by sublimator** [comment]
squashed

**Comment by indutny** [comment]
Landed and released, thank you!

**Comment by sublimator** [comment]
cheers :)

**Pull request closed without merging**